### ISSUE TITLE
C# (CSharp)  interface usage for joint prediction (search)

### ISSUE THREAD

**Issue by andy-soft** [VowpalWabbit/vowpal_wabbit#2524]
### Description

> A brief description of the error, missing documentation or what you would like added
I want to use VW from C# and not use the command line, just use a service and feed data in (to train), and pump predictions out 
The problem is I did not find the C#API, no description of the usage, the Wiki info and the readme are not enough to understand the interfacing models. 

I wat to use search, and do POS tagging, NER and Dependency Parsing, (all sequential) and don't know how to feed the appropriated formatted strings and get the prediction out, the internal classes are complicated to inference from!

I successfully got the multiclass classification to work 
but BUG => the linked class (the DLL) don't get released (the operating-system file-lock) even after disposing the VW object and removing the whole assembly! So there is a serious threat to put this into a windows service!

Please help!

### Link to Documentation Page

> Where is the documentation in question?

**Comment by lokitoth** [comment]
I added a link to an [example gist](lokitoth/bd8e9e6f1c1be57c0cc320e638ce6b29) in #2578. Will use this issue to track the work to provide better documentation, and focus on assisting @andy-soft with specific questions in the other one.

**Comment by andy-soft** [comment]
Thank you!
I am testing it right now! (and it seems to work!)
But.. I found no information on the documentation to allow me do this construct!!!
Thanks a lot!
Just I am testing the NER parser (in Spanish) and it overfits a lot, it's really fast but commits lots of errors!
F1 < 60% when others reach 92% on same dataset. (Ancora) I also added lots of features on the peature space, not only word and POS

**Comment by andy-soft** [comment]
Hi there, it worked fine! 
There are some issues out there, 
¿may I can get some additional help?

I Want to make use form C# of the reinforcement and online learning features, 
For this, to work fine I need the ability to create in C# the functions for the GOLD-Standard (Oracle) to be useful, and even to train online, but in Python there are plenty of samples of the function, but unfortunately under .NET (C# interface API)  I did not manage hoy to link this function, just because my joint-prediction will be context-aware (I want to parse natural language using a dynamic semantic-context) I am a heavy NLP researcher but unfortunately did not find hoy to build this linking.
¿Can you provide me a sample like the other, which I found really enlightening?
I am managing to use the C# interface for training even (multiline) backed out of what you sent me, when ready I'll share it with you.

VW quality CLAIMS on POS-Tagging and NER / NEC extraction are somehow FAKE or not entirely true.

Last but not least, the claimed on most papers (around 2016) F-Score of the NER and POS tagger are false.
I tested it with C#, and the SAME Ancora dataset, and the resulting score is not true, (F-Score it's far lower, around 60%)
The library DO NOT GENERALIZE WELL in SPANISH, and despite the apparent scores of the training are high, the use with other data fails, and the recognition of tags is poor, no to say bad!
¿Maybe the library is overfitting too much? 

I guess I am training well, on a well-converted dataset towards the vw format, and the loss goes really down!
Buth the predictions on data outside the dataset are pretty bad!
Tried lots of parameter, cubic, and cuadratic relation on multiple namespaces, none helped!
¿Are there  a way to measure the overfitting capability, apart from the loss factor provided)

is there a real POS-Tagger or NER/NEC extractor with an example to reproduce (including dataset) to see what may I be doing wrong?

thanks anyway!

**Comment by JohnLangford** [comment]
Closing this now as obsolete.  Please reopen if it's still relevant.
